python/social_bot/envs/grocery_ground.py

```diff
@@ -0,0 +1,232 @@
+
+
+class GroceryGround(gym.Env):
+    """
+    The goal of this task is to train the agent to navigate to the objects given its
+    name.
+
+    Joints of caster and arms in the PR2 are controllable by force,
+    the observations are the states of the world, including every model's position
+    and rotation.
+
+    In the version:
+    * agent will receive the total contacts of the wall as the reward
+
+    """
+
+    def __init__(self, with_language=False, use_image_obs=False, port=None):
+        """
+        Args:
+            with_language (bool): the observation will be a dict with an extra sentence
+            use_image_obs (bool): use image as observation, or use pose of the objects
+            port: Gazebo port, need to specify when run multiple environment in parallel
+        """
+        if port is None:
+            port = 0
+        gazebo.initialize(port=port)
+        self._world = gazebo.new_world_from_file(
+            os.path.join(social_bot.get_world_dir(), "grocery_ground.world"))
+        # self._world.info()
+
+        self._agent = self._world.get_agent()
+        self._pr2_joints = list(
+            filter(
+                lambda s: s.find('fl_caster_r_wheel_joint') != -1 or s.find('fr_caster_l_wheel_joint') != -1,
+                self._agent.get_joint_names()))
+        logger.info("joints to control: %s" % self._pr2_joints)
+        self._objects_names = [
+            'coke_can_0', 'coke_can_1', 'coke_can_2', 'coke_can_3',
+            'coke_can_4', 'table', 'grey_wall', 'cube_20k', 'caffe_table',
+            'bookshelf', 'car_wheel_0', 'car_wheel_1', 'trash_can',
+            'plastic_cup_0', 'plastic_cup_1', 'plastic_cup_2', 'plastic_cup_3',
+            'plastic_cup_4'
+        ]
+        self._goal_name = 'bookshelf'
+
+        self._teacher = teacher.Teacher(False)
+        task_group = teacher.TaskGroup()
+        task_group.add_task(GroceryGroundGoalTask())
+        self._teacher.add_task_group(task_group)
+
+        self._with_language = with_language
+        self._use_image_obs = use_image_obs
+        self._resized_image_size = (84, 84)
+
+        obs = self.reset()
+        if self._use_image_obs:
+            obs_data_space = gym.spaces.Box(
+                low=0, high=1, shape=obs.shape, dtype=np.uint8)
+        else:
+            obs_data_space = gym.spaces.Box(
+                low=-50, high=50, shape=obs.shape, dtype=np.float32)
+        control_space = gym.spaces.Box(
+            low=-20, high=20, shape=[len(self._pr2_joints)], dtype=np.float32)
+
+        if self._with_language:
+            self.observation_space = gym.spaces.Dict(
+                data=obs_data_space, sentence=DiscreteSequence(128, 24))
+            self._action_space = gym.spaces.Dict(
+                control=control_space, sentence=DiscreteSequence(128, 24))
+        else:
+            self.observation_space = obs_data_space
+            self.action_space = control_space
+
+    def reset(self):
```

Comment:
Let's create a random map for each episode. We can control how many types of objects are in one map. This should be controlled by Teacher so that teacher can control the task difficulty according to the progress of the agent.

Reply:
And we should make it easier to switch agent (pr2, pioneer, etc)